src/BotPlutusInterface/Balance.hs

```diff
@@ -83,16 +81,45 @@ preBalanceTxIO pabConf ownPkh unbalancedTx =
 
       lift $ printLog @w Debug $ show utxoIndex
 
-      loop utxoIndex privKeys requiredSigs [] tx
+      -- Adds required collaterals, only needs to happen once
+      -- Also adds signatures for fee calculation
+      preBalancedTx <- hoistEither $ addTxCollaterals utxoIndex tx >>= addSignatories ownPkh privKeys requiredSigs
+
+      -- Balance the tx
+      (balancedTx, minUtxos) <- loop utxoIndex privKeys [] preBalancedTx
+
+      -- Check if we have Ada change
+      let adaChange = getAdaChange utxoIndex (lovelaceValue $ txFee balancedTx) balancedTx
+      -- If we have no change UTxO, but we do have change, we need to add an output for it
+      -- We'll add a minimal output, run the loop again so it gets minUTxO, then update change
+      balancedTxWithChange <-
+        if adaChange /= 0 && not (hasChangeUTxO ownPkh balancedTx)
```

Comment:
It's got a bit complicated here, I wonder if we could somehow refactor this. But for that, maybe it's best if we create some property tests for this.. which could be a bit laborious as we we need to simulate the calculateFee and calculateMinUtxo behaviours somehow, so it might be best if we do it in a later PR.

Reply:
I'll try refactor it a bit, but yeah, the tests are a bit lacking, simply due to the intended behavior being complex and somewhat hidden